File: probs/cses/1739/1739.cpp
#include <bits/stdc++.h>

using namespace std;

const int MAXN = 1005;

int n, bit[MAXN][MAXN];

void updy(int x, int y, int val)
{
	while(y <= n)
	{
		bit[x][y] += val;
		y += (y&-y);
	}
}

void upd(int x, int y, int val)
{
	while(x <= n)
	{
		updy(x, y, val);
		x += (x&-x);
	}
}

int quey(int x, int y)
{
	int ret = 0;
	while(y > 0)
	{
		ret += bit[x][y];
		y -= (y & -y);
	}
	return ret;
}

int que(int x, int y)
{
	int ret = 0;
	while(x > 0)
	{
		ret += quey(x, y);
		x -= (x & -x);
	}
	return ret;
}

int que(int x0, int y0, int x1, int y1)
{
	int b1 = que(max(x0, x1), max(y0, y1));
	int b2 = que(min(x0, x1)-1, max(y0, y1));
	int b3 = que(max(x0, x1), min(y0, y1)-1);
	int b4 = que(min(x0, x1)-1, min(y0, y1)-1);
	return b1-b2-b3+b4;
}

bool has[MAXN][MAXN];

int main()
{
	int t;
	cin >> n >> t;
	for(int i = 1; i <= n; i++) for(int j = 1; j <= n; j++)
	{
		char k;
		cin >> k;
		if(k == '*') upd(i, j, 1);
		has[i][j] = (k=='*');
	}
	while(t--)
	{
		int ty; cin >> ty;
		if(ty == 1)
		{
			int x, y; cin >> x >> y;
			upd(x, y, (has[x][y] ? -1 : 1));
			has[x][y] = !has[x][y];
		} else 
		{
			int x0, x1, y0, y1;
			cin >> x0 >> y0 >> x1 >> y1;
			cout << que(x0, y0, x1, y1) << endl;
		}
	}
}

File: probs/cses/1739/1739.py
# 2d bit but in python

n, q = map(int,input().split())

s = []
for i in range(n):
    s.append(input())


bit = [ [0 for i in range(n+5)] for i in range(n+5) ]

def updy(x, y, val):
    while y <= n:
        bit[x][y] += val 
        y += (y&-y)

def upd(x, y, val):
    while x <= n:
        updy(x, y, val)
        x += (x&-x)

def quey(x, y):
    ret = 0
    while y > 0: 
        ret += bit[x][y]
        y -= (y&-y)
    return ret 

def que(x, y):
    ret = 0
    while x > 0:
        ret += quey(x, y)
        x -= (x&-x)
    return ret 

def qu(x0, y0, x1, y1):
    if x1 < x0: 
        p = x1
        x1 = x0
        x0 = p 
    if y1 < y0:
        p = y1
        y1 = y0 
        y0 = p 
    x0 -= 1
    y0 -= 1
    b1 = que(x1, y1)
    b2 = que(x0, y1)
    b3 = que(x1, y0)
    b4 = que(x0, y0)
    return b1-b2-b3+b4 

A = [ [ 0 for i in range(n) ] for i in range(n) ]
for i in range(n):
    for j in range(n):
        if s[i][j] == '*':
            upd(i+1, j+1, 1)
            A[i][j] = 1

for i in range(q):
    inp = list(map(int, input().split()))
    if inp[0] == 1:
        x = inp[1]
        y = inp[2]
        upd(x, y, (-1 if A[x-1][y-1] == 1 else 1))
        A[x-1][y-1] = 1-A[x-1][y-1]
    else:
        x0 = inp[1]
        y0 = inp[2]
        x1 = inp[3]
        y1 = inp[4]
        print(qu(x0, y0, x1, y1))